### PR TITLE
Adds comment placeholders

### PR DESCRIPTION
**Pull request by magikid**
Finally doing a full code review of your project.  I mostly stuck to the JS since it was a JS class but I also had some thoughts on your PHP code since that's what I use day-to-day.

### CODE REVIEW

File: model/userlogin.php [hunk 1]
```diff
@@ -8,12 +8,12 @@ class Users
 
   private $conn;  //Database connector
 
-    function __construct( $user=false, $pass=false ) 
+    function __construct( $user=false, $pass=false ) // Types for paramters
```
Comment (magikid) [inline model/userlogin.php:11]: Instead of using `false` as a default value, you should use types like `function __construct(string $user, string $pass)`.  That way you don't need to check that they're set.  You would then check that they're non-zero length `mb_strlen($user)>0`.

File: model/userlogin.php [hunk 2]
```diff
@@ -8,12 +8,12 @@ class Users
...
     {
         $this->user 	= ( isset( $user ) && $user != '' && $user != false ? $user: false );
         $this->pass 	= ( isset( $pass ) && $pass != '' && $pass != false ? $pass: false );
         $this->error 	= false;
-        $this->connect();
+        $this->connect(); // Constructor doing work
```
Comment (magikid) [inline model/userlogin.php:16]: You shouldn't do work in the constructor.  This should be done by the caller.

File: model/userlogin.php [hunk 3]
```diff
@@ -24,20 +24,20 @@ function connect()
 
 	}
 
-    function Login($user, $pass)
+    function Login($user, $pass) // Allows user to control flow
     {
 
       $sql = " SELECT id, firstname, lastname, password ".
-             " FROM users where username = '$user';";
+             " FROM users where username = '$user';"; // Yikes! SQL injection possible, use PDO#bindParam
```
Comment (magikid) [inline model/userlogin.php:31]: You should never be inlining a string into SQL like this.  A specially crafted string could cause an SQL injection attack right here.  Instead, you should be [using bind params](https://www.php.net/manual/en/pdostatement.bindparam.php#refsect1-pdostatement.bindparam-examples)

File: model/userlogin.php [hunk 4]
```diff
@@ -24,20 +24,20 @@ function connect()
 
 	}
 
-    function Login($user, $pass)
+    function Login($user, $pass) // Allows user to control flow
```
Comment (magikid) [inline model/userlogin.php:27]: The server should be responding to either grant access or deny it.  The way this is currently written, a crafty user could access the protected area without properly logging in.

File: model/userlogin.php [hunk 5]
```diff
@@ -24,20 +24,20 @@ function connect()
...
     {
 
       $sql = " SELECT id, firstname, lastname, password ".
-             " FROM users where username = '$user';";
+             " FROM users where username = '$user';"; // Yikes! SQL injection possible, use PDO#bindParam
 
       $ret = $this-> conn ->query($sql);
-      $result =$ret->fetchAll();
+      $result =$ret->fetchAll();  // Why fetchAll for a single result?
```
Comment (magikid) [inline model/userlogin.php:34]: Why fetchAll for a single result?

File: model/userlogin.php [hunk 6]
```diff
@@ -24,20 +24,20 @@ function connect()
...
 
       if (count($result) > 0) 
       {
             foreach ($result as $Record) 
             {
-                if($Record['password'] === $pass)
+                if($Record['password'] === $pass) // The password should *never* be stored in plain text
```
Comment (magikid) [inline model/userlogin.php:40]: Yikes! Storing passwords in plain text in the DB is a serious no-no.  Anyone gaining access like from the SQL injection attack above would be able to log in as anyone.

File: model/userlogin.php [hunk 7]
```diff
@@ -47,7 +47,7 @@ function Login($user, $pass)
                     return  $Record; // Wrong password.
                 }
             }
-        } else 
+        } else // Funky spacing
```
Comment (magikid) [inline model/userlogin.php:50]: You have different levels of spacing throughout the entire project.  This makes it really hard to follow where everything begins and ends.  It's technically not a problem but it's a [code smell](https://en.wikipedia.org/wiki/Code_smell)l.

File: model/userlogin.php [hunk 8]
```diff
@@ -62,12 +62,12 @@ function Login($user, $pass)
     return;
   }
 
-  $userName =   htmlspecialchars($_GET['username']);
-  $UserPassword =  htmlspecialchars($_GET['userpassword']);
- 
+  $userName =   htmlspecialchars($_GET['username']); // Not how to handle characters for SQL
```
Comment (magikid) [inline model/userlogin.php:65]: I think I've already covered it above but this isn't how you should be handling special characters for SQL.

File: model/userlogin.php [hunk 9]
```diff
@@ -62,12 +62,12 @@ function Login($user, $pass)
...
+  $UserPassword =  htmlspecialchars($_GET['userpassword']); // Inconsistent capitalization
```
Comment (magikid) [inline model/userlogin.php:66]: Throughout the project, you have random capitalization on your variables.  Again, technically this isn't a problem but it makes it really hard to read your code.  It's another code smell.

File: model/userlogin.php [hunk 10]
```diff
@@ -62,12 +62,12 @@ function Login($user, $pass)
...
+
   $objUser = new Users();
   $logUser = $objUser->Login( $userName, $UserPassword );
 
-  echo json_encode( $logUser );
+  echo json_encode( $logUser ); // This also returns their password?!
```
Comment (magikid) [inline model/userlogin.php:71]: Why are you returning the entire object here?

**Comment by jarvyii** [comment]
Hi Chris,

Thank you so much for your time helping me with my code.
I appreciate it a lot. It has been very useful for me.
My thanks for your dedication to teaching us.



Jose A Reynaldo
________________________________
From: Chris Jones ***@***.***>
Sent: Tuesday, June 1, 2021 11:15 AM
To: jarvyii/Jobsclocking ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [jarvyii/Jobsclocking] Adds comment placeholders (#1)


Finally doing a full code review of your project. I mostly stuck to the JS since it was a JS class but I also had some thoughts on your PHP code since that's what I use day-to-day.

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/jarvyii/Jobsclocking/pull/1

Commit Summary

  *   Adds comment placeholders

File Changes

  *   M model/userlogin.php<https://github.com/jarvyii/Jobsclocking/pull/1/files#diff-7b083de5d647a404cc413205b0ef0bd8413bb36dc5ebab5e5d92f218e2b26392> (22)
  *   M view/login.js<https://github.com/jarvyii/Jobsclocking/pull/1/files#diff-abd6fccddbce6bffc01076d6417b6c6d6c81435ff4672421583e862b624175e4> (28)
  *   M view/timecard.js<https://github.com/jarvyii/Jobsclocking/pull/1/files#diff-6495bcc4fd632b761b766cebae78c8e4ef17c41730568cc1bd6ad321a976922e> (14)

Patch Links:

  *   https://github.com/jarvyii/Jobsclocking/pull/1.patch
  *   https://github.com/jarvyii/Jobsclocking/pull/1.diff

File: view/login.js [hunk 11]
```diff
@@ -1,9 +1,9 @@
 
-function createDiv(Type, Id, className, Text)
+function createDiv(Type, Id, className, Text) // Inconsistent variable capitalization
 {
-   let Div = document.createElement( Type ); 
+   let Div = document.createElement( Type ); // Inconsistent spacing
```
Comment (magikid) [inline view/login.js:4]: Why do you have spaces around `Type` here and not anywhere else?  This is another code smell.

File: view/login.js [hunk 12]
```diff
@@ -1,9 +1,9 @@
...
 
-   if( Id != "" ) {
+   if( Id != "" ) { // Using double equals
```
Comment (magikid) [inline view/login.js:6]: Like in PHP, you should be using double equals here so that both value and type are checked.  `if(Id!=="")`

File: view/login.js [hunk 13]
```diff
@@ -21,7 +21,7 @@ function createDiv(Type, Id, className, Text)
 
 function createForm() {
     const form = document.createElement("form");
-    form.setAttribute("method", "GET"); 
+    form.setAttribute("method", "GET");  // Using GET for login
```
Comment (magikid) [inline view/login.js:24]: You should *never* use a `GET` request for a login form because all of the form data will be visible in the URL to anyone (malicious or otherwise) who gets their hands on it.  A GET request might also be cached whereas a POST request won't be.

File: view/login.js [hunk 14]
```diff
@@ -34,10 +34,10 @@ function  createLoginForm() {
     form.appendChild(divForm);
     $("#divlogin").append(form);
 
-    const divImage = '<div class="imgcontainer"><img src="img/login.jpg" alt="Login now" class="avatar"></div>';
-    document.getElementById("divForm").innerHTML = divImage;  
+    const divImage = '<div class="imgcontainer"><img src="img/login.jpg" alt="Login now" class="avatar"></div>'; // divImage only used here
```
Comment (magikid) [inline view/login.js:37]: For all of these below, you assign a string to a variable and then immediately assign that variable to the `innerHTML` and never use the variable again.  All these `const` assignments should be removed.

File: view/login.js [hunk 15]
```diff
@@ -34,10 +34,10 @@ function  createLoginForm() {
...
+    document.getElementById("divForm").innerHTML = divImage;
 
-    const divloginimput= createDiv("div", "loginimput", "loginimput", "");
+    const divloginimput= createDiv("div", "loginimput", "loginimput", ""); // Typo in variable name
```
Comment (magikid) [inline view/login.js:40]: You have a typo in several of these variable names.  You have used i**m**put instead of i**n**put.

File: view/login.js [hunk 16]
```diff
@@ -48,7 +48,7 @@ function  createLoginForm() {
 
     let user = document.getElementById("user-name" );
     //user.title = "Pattern = Email Address";
-    user.pattern = "^[^\s@]+@([^\s@.,]+\.)+[^\s@.,]{2,}$";
+    user.pattern = "^[^\s@]+@([^\s@.,]+\.)+[^\s@.,]{2,}$"; // Seems unused
```
Comment (magikid) [inline view/login.js:51]: You have like three different regexes scattered throughout the codebase for validating email addresses.  There should be only one.  Also, why are you storing this in the DOM?

File: view/login.js [hunk 17]
```diff
@@ -73,6 +73,7 @@ function  createLoginForm() {
 
 }
 
+// Unnecessary function
```
Comment (magikid) [inline view/login.js:76]: Why did you write your own function when you are also using jQuery?  You should be using the `$.get()` function instead of rolling your own.

File: view/login.js [hunk 18]
```diff
@@ -132,12 +133,12 @@ function checkUser(){
 
     ajaxGET( myURL )
          .then( validUser )
-         .catch( err => alert( err ) );
+         .catch( err => alert( err ) ); // Using alert for error
```
Comment (magikid) [inline view/login.js:136]: Using alert is a bad practice.  If an error occurs here, the user is unlikely to understand whatever error message is spit out to them.  This should probably be `console.log` instead.

File: view/login.js [hunk 19]
```diff
@@ -132,12 +133,12 @@ function checkUser(){
...
 
 
 
 }
-function tovalidEmail() {
+function tovalidEmail() { // Unused function
```
Comment (magikid) [inline view/login.js:141]: I don't think this is ever called.  Unused code should be deleted.  Once it's been in a single commit, git will always remember it if you ever need to go back to it.

File: view/login.js [hunk 20]
```diff
@@ -162,8 +163,8 @@ function showOrHideTip(show, element) {
   
 }
 function createListener(validator) {
-  return e => {
-    
+  return e => { // e is never defined,
```
Comment (magikid) [inline view/login.js:166]: This is such an unusual way to write this that I just now realized this `createListener` returns a function that accepts a single parameter of `e`.  This seems like a tad overengineered since it's called in a single place.

File: view/login.js [hunk 21]
```diff
@@ -174,7 +175,7 @@ function createListener(validator) {
     showOrHideTip(showTip, tooltip);
   };
 }
-
+    // Funky indentation
```
Comment (magikid) [inline view/login.js:178]: The following block is indented as though it is inside a function but it's not.  It runs when the JS is parsed and run.

File: view/login.js [hunk 22]
```diff
@@ -195,6 +196,7 @@ function createListener(validator) {
             } );
  
 
+// Unused block
```
Comment (magikid) [inline view/login.js:199]: Unused code should be deleted.

File: view/timecard.js [hunk 23]
```diff
@@ -17,7 +17,7 @@ const totalWayPayments = Description.length ;
 function writePaymentsDescription(){
    
       function addDescrption(value){    //Add the description to the column
-          var p = document.createElement("INPUT");
+          var p = document.createElement("INPUT"); // Capitalized
```
Comment (magikid) [inline view/timecard.js:20]: Why is `INPUT` all caps here?

File: view/timecard.js [hunk 24]
```diff
@@ -180,12 +180,12 @@ function startOfWeek(date)
...
   {
-     
+
     var lastday = date.getDate() - (date.getDay() - 1) + 6;
     return new Date(date.setDate(lastday));
- 
+
   }
-function createFormElement( name, value, type , status){
+function createFormElement( name, value, type , status){ // Duplicated logic with other JS?
```
Comment (magikid) [inline view/timecard.js:188]: This is almost the same as in the other file.  Close enough that there should only be a single function to handle both cases.

File: view/timecard.js [hunk 25]
```diff
@@ -265,8 +265,8 @@ function createFormTimecard(User){
   /*******************************************
  Call a (PHP) Backend function to save data with the specific URL.
 ********************************************/
-function postAJAX( Url, formClass ){
-    
+function postAJAX( Url, formClass ){ // Unneeded function
```
Comment (magikid) [inline view/timecard.js:268]: Again, this is provided by jQuery and you shouldn't roll your own `$.post()`.

File: view/timecard.js [hunk 26]
```diff
@@ -310,7 +310,7 @@ $('#save').click(function () {
 
    }); // /$('#save').click(function ());
 
-  $(document).ready(  function() {
+  $(document).ready(  function() { // Unneeded function
```
Comment (magikid) [inline view/timecard.js:313]: Unused code should be deleted.
